### PR TITLE
Introduce `__future__.annotations` to TPE-related modules

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

Relates to:
- https://github.com/optuna/optuna/issues/4508

## Description of the changes

Introduce `__future__.annotations` to TPE-related modules.

### CODE REVIEW

**Comment by nabenabe0928** [comment]
@y0z @contramundum53 Could you review this PR?

File: optuna/samplers/_tpe/probability_distributions.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
-from typing import List
+from __future__ import annotations
+
 from typing import NamedTuple
 from typing import Union
```
Comment (eukaryo) [inline optuna/samplers/_tpe/probability_distributions.py:4]: Could we remove this line and change the code using `Union` as following? (I confirmed that this change works in local python3.11 environment...)
```
_BatchedDistributions = (
    _BatchedCategoricalDistributions
    | _BatchedTruncNormDistributions
    | _BatchedDiscreteTruncNormDistributions
)
```

Comment (nabenabe0928) [reply optuna/samplers/_tpe/probability_distributions.py:4]: I can check whether the check passes with this modification, but if I remember correctly, it does not work.

Comment (nabenabe0928) [reply optuna/samplers/_tpe/probability_distributions.py]: It did not work, so I will revert the change!
https://github.com/optuna/optuna/actions/runs/9059459819/job/24887097949?pr=5443

Comment (eukaryo) [reply optuna/samplers/_tpe/probability_distributions.py:4]: Thank you for trying it!

**Review by eukaryo — APPROVED** [review]
LGTM

**Review by y0z — APPROVED** [review]
LGTM

**Pull request merged**